### ISSUE TITLE
VercelPostgresError - 'invalid_connection_string' when using createPool

### ISSUE THREAD

**Issue by ievenight** [saasfly/saasfly#73]
## Description:
when I try to `bun run dev:web` and open `http://localhost:3000/en`, something went wrong as the picture.

VercelPostgresError: VercelPostgresError - 'invalid_connection_string': This connection string is meant to be used with a direct connection. Make sure to use a pooled connection string or try `createClient()` instead.

## Error Details:
VercelPostgresError: VercelPostgresError - 'invalid_connection_string': This connection string is meant to be used with a direct connection. Make sure to use a pooled connection string or try `createClient()` instead.

<img width="882" alt="image" src="https://github.com/saasfly/saasfly/assets/54346544/025ef66c-c4a3-436f-9760-66d1e140b7bd">

## Additional Context:

I have added postgres_url to .env.local, `
POSTGRES_URL='postgres://default:<xxxx>@<xxxx>.aws.neon.tech:5432/verceldb?sslmode=require'`

and `bun db:push` successfully

**Comment by semlinker** [comment]
I am using a local PostgreSQL server and have set the POSTGRES_URL to "postgres://postgres:123456@127.0.0.1:5432/saasfly", but it still throws an "invalid_connection_string" error. Should it be configured according to the following article? —— https://gal.hagever.com/posts/running-vercel-postgres-locally

**Comment by jackc3** [comment]
@vercel/postgres-kysely
This is a Serverless driver. You need apply a Postgres on Vercel

Or you can change the driver:

export const db = new Kysely<DB>({
  dialect: new PostgresDialect({
    pool: new Pool({
      connectionString: process.env.DATABASE_URL,
    }),
  }),
});

<img width="1885" alt="image" src="https://github.com/saasfly/saasfly/assets/10334353/0e83dbfc-9568-4b3a-a9e2-8d0a7b3bc571">

**Comment by semlinker** [comment]
@jackc3 Thanks, it works.
`
import { Pool } from 'pg'
import { Kysely, PostgresDialect } from "kysely"

export const db = new Kysely<Database>({
  dialect: new PostgresDialect({
    pool: new Pool({
      connectionString: "postgres://postgres:123456@127.0.0.1:5432/saasfly",
    }),
  }),
});
`

**Comment by Steravy** [comment]
Hello. Same problem here but the solution provided did not fix it.

**Comment by Steravy** [comment]
![Screenshot from 2024-06-25 22-11-43](https://github.com/saasfly/saasfly/assets/116315400/9e2eccf0-8e7e-4a3b-a55b-6ece370ed019)

**Comment by jackc3** [comment]
<img width="428" alt="image" src="https://github.com/saasfly/saasfly/assets/10334353/d01ba800-ecec-40d8-ae19-db72973cb1db">

please insure the host can be connected and `any-db` was created

**Comment by Steravy** [comment]
@jackc3 i am assuming it can be connected because when i run `bun db:push` the changes was made in the database but when i started the development server the error is thrown. The connection string i put there is an example.

**Comment by Steravy** [comment]
> > @jackc3 i am assuming it can be connected because when i run `bun db:push` the changes was made in the database but when i started the development server the error is thrown. The connection string i put there is an example.
> 
> do you change your package.json ? "dependencies": { "kysely": "0.27.2", "pg": "8.11.3", "@t3-oss/env-nextjs": "0.7.3" },

No, i didn`t. Should i?

**Comment by Steravy** [comment]
> > > > @jackc3 i am assuming it can be connected because when i run `bun db:push` the changes was made in the database but when i started the development server the error is thrown. The connection string i put there is an example.
> > > 
> > > 
> > > do you change your package.json ? "dependencies": { "kysely": "0.27.2", "pg": "8.11.3", "@t3-oss/env-nextjs": "0.7.3" },
> > 
> > 
> > No, i didn`t. Should i?
> > yes , you need to remove  "@vercel/postgres-kysely": "0.8.0" and use pg": "8.11.3". but note, this will cause problems in the Serveless server.

Not planing in deploying it on vercel. But any suggestion on the possible problems or ones that you have faced?

**Comment by Steravy** [comment]
> > > > > > @jackc3 i am assuming it can be connected because when i run `bun db:push` the changes was made in the database but when i started the development server the error is thrown. The connection string i put there is an example.
> > > > > 
> > > > > 
> > > > > do you change your package.json ? "dependencies": { "kysely": "0.27.2", "pg": "8.11.3", "@t3-oss/env-nextjs": "0.7.3" },
> > > > 
> > > > 
> > > > No, i didn`t. Should i?
> > > > yes , you need to remove  "@vercel/postgres-kysely": "0.8.0" and use pg": "8.11.3". but note, this will cause problems in the Serveless server.
> > 
> > 
> > Not planing in deploying it on vercel. But any suggestion on the possible problems or ones that you have faced?
> > Deploy to a VPS looks like the simplest method.

Yeah that is what i was planing to do. I will make the changes. Thank you for your time @tianzx

**Comment by Sheikh-Srijon** [comment]
Hi, I'm using Supabase pooling url but keep getting invalid_connection_string. 

Can you show me the format of you bun run build command in ci.yml file? 

![image](https://github.com/saasfly/saasfly/assets/47705083/cc1f1919-2410-47aa-8563-a84b306d7640)

I have tried using both direct and pooler string from supabase. Same error

**Comment by jackc3** [comment]
Run it locally, you should:
1. change your package.json(s)
`"dependencies": {
"kysely": "0.27.2",
"pg": "8.11.3",
"@t3-oss/env-nextjs": "0.7.3"
}`
2. change db/index.ts
`import { Pool } from 'pg'
import { Kysely, PostgresDialect } from "kysely"

export const db = new Kysely<Database>({
  dialect: new PostgresDialect({
    pool: new Pool({
      connectionString: "postgres://<your_username>:<your_password>@<host>:<port>/<database>",
    }),
  }),
});`

**Comment by jackc3** [comment]
![Uploading image.png…]()
@Sheikh-Srijon

**Comment by jackc3** [comment]
<img width="843" alt="image" src="https://github.com/saasfly/saasfly/assets/10334353/c62fe6b3-3c4c-43d7-9b06-d6ae1d3376a1">
@Sheikh-Srijon

**Comment by jackc3** [comment]
Have you changed package.json(s)?
`"dependencies": { "kysely": "0.27.2", "pg": "8.11.3", "@t3-oss/env-nextjs": "0.7.3" }`

@shuther

**Comment by jackc3** [comment]
> Thanks, but not sure about the next steps. Should I also modify auth/db.ts?

@shuther yes, you should modify it

**Comment by jkeck99** [comment]
Hi @tianzx.  I'm wondering if it is possible to to keep "@vercel/postgres-kysely": "0.8.0" in package.json for serverless deployments (preview & production) while also using adding necessary dependencies for local development "dependencies": { "kysely": "0.27.2", "pg": "8.11.3", "@t3-oss/env-nextjs": "0.7.3" } and using a different driver? export const db = new Kysely<DB>({
  dialect: new PostgresDialect({
    pool: new Pool({
      connectionString: process.env.DATABASE_URL,
    }),
  }),
});

If that's possible, I can conditionally load the right DB based on the environment.  If not, I'll have to use serverless environment for local development which is not ideal. Thanks for any advice.
